coderd/coderdtest/authorize.go

```diff
@@ -1,559 +1,144 @@
-
-		// These endpoints need payloads to get to the auth part. Payloads will be required
-		"PUT:/api/v2/users/{user}/roles":                                {StatusCode: http.StatusBadRequest, NoAuthorize: true},
-		"PUT:/api/v2/organizations/{organization}/members/{user}/roles": {NoAuthorize: true},
-		"POST:/api/v2/workspaces/{workspace}/builds":                    {StatusCode: http.StatusBadRequest, NoAuthorize: true},
-		"POST:/api/v2/organizations/{organization}/templateversions":    {StatusCode: http.StatusBadRequest, NoAuthorize: true},
-
-		// For any route using SQL filters, we do not check authorization.
-		// This is because the in memory fake does not use SQL.
-		"GET:/api/v2/workspaces/": {
-			StatusCode:   http.StatusOK,
-			NoAuthorize:  true,
-			AssertAction: rbac.ActionRead,
-			AssertObject: rbac.ResourceWorkspace,
-		},
-		"GET:/api/v2/organizations/{organization}/templates": {
-			StatusCode:   http.StatusOK,
-			NoAuthorize:  true,
-			AssertAction: rbac.ActionRead,
-			AssertObject: rbac.ResourceTemplate,
-		},
+	Recorder *RecordingAuthorizer
+}
 
-		"GET:/api/v2/debug/coordinator": {
-			AssertAction: rbac.ActionRead,
-			AssertObject: rbac.ResourceDebugInfo,
-		},
+// AssertRBAC returns an RBACAsserter for the given user. This asserter will
+// allow asserting that the correct RBAC checks are performed for the given user.
+// All checks that are not run against this user will be ignored.
+func AssertRBAC(t *testing.T, api *coderd.API, client *codersdk.Client) RBACAsserter {
+	if client.SessionToken() == "" {
+		t.Fatal("client must be logged in")
 	}
-
-	// Routes like proxy routes support all HTTP methods. A helper func to expand
-	// 1 url to all http methods.
-	assertAllHTTPMethods := func(url string, check RouteCheck) {
-		methods := []string{
-			http.MethodGet, http.MethodHead, http.MethodPost,
-			http.MethodPut, http.MethodPatch, http.MethodDelete,
-			http.MethodConnect, http.MethodOptions, http.MethodTrace,
-		}
-
-		for _, method := range methods {
-			route := method + ":" + url
-			assertRoute[route] = check
-		}
+	recorder, ok := api.Authorizer.(*RecordingAuthorizer)
+	if !ok {
+		t.Fatal("expected RecordingAuthorizer")
 	}
 
-	assertAllHTTPMethods("/%40{user}/{workspace_and_agent}/apps/{workspaceapp}/*", RouteCheck{
-		AssertAction: rbac.ActionCreate,
-		AssertObject: applicationConnectObj,
-	})
-	assertAllHTTPMethods("/@{user}/{workspace_and_agent}/apps/{workspaceapp}/*", RouteCheck{
-		AssertAction: rbac.ActionCreate,
-		AssertObject: applicationConnectObj,
-	})
-
-	return skipRoutes, assertRoute
-}
-
-type RouteCheck struct {
-	NoAuthorize  bool
-	AssertAction rbac.Action
-	AssertObject rbac.Object
-	StatusCode   int
+	// We use the database directly to not cause additional auth checks on behalf
+	// of the user. This does add authz checks on behalf of the system user, but
+	// it is hard to avoid that.
+	ctx := dbauthz.AsSystemRestricted(context.Background())
```

Comment:
It works! 

![image](https://user-images.githubusercontent.com/4949514/226332398-e597339f-2102-4242-a585-91776cca4eba.png)